comms/dht/src/network_discovery/seed_strap.rs

```diff
@@ -156,326 +158,97 @@ impl SeedStrap {
-                continue;
-            }
-
-            debug!(
-                target: LOG_TARGET,
-                "SeedStrap: Iteration {}/{}: Attempting to connect to seed peer '{}' to get their peer list",
-                idx + 1,
-                num_seeds_to_try,
-                seed_peer_node_id_str
-            );
-
-            let mut conn = match self
-                .context
-                .connectivity
-                .dial_peer(seed_peer_candidate.node_id.clone())
-                .await
-            {
-                Ok(c) => c,
-                Err(e) => {
-                    warn!(
-                        target: LOG_TARGET,
-                        "SeedStrap: Iteration {}/{}: Failed to dial seed peer '{}': {}. Continuing to next seed candidate.",
-                        idx + 1,
-                        num_seeds_to_try,
-                        seed_peer_node_id_str,
-                        e
-                    );
-                    continue;
-                },
-            };
-
-            debug!(
-                target: LOG_TARGET,
-                "SeedStrap: Connected to seed peer '{}'. Requesting peer list.",
-                seed_peer_node_id_str
-            );
-
-            let peers_from_seed = match self.fetch_peers_from_connection(&mut conn).await {
-                Ok(peers) => peers,
-                Err(e) => {
-                    warn!(
-                        target: LOG_TARGET,
-                        "SeedStrap: Iteration {}/{}: Failed to fetch peers from seed peer '{}': {}. Disconnecting and continuing.",
-                        idx + 1,
-                        num_seeds_to_try,
-                        seed_peer_node_id_str,
-                        e
-                    );
-                    // Attempt to disconnect, log if it fails but continue
-                    if let Err(disc_err) = conn.disconnect(Minimized::Yes).await {
-                        warn!(target: LOG_TARGET, "SeedStrap: Also failed to disconnect from seed peer '{}' after fetch failure: {}", seed_peer_node_id_str, disc_err);
-                    }
-                    continue;
-                },
-            };
-
-            if peers_from_seed.is_empty() {
-                info!(
+        let mut seed_peers = seed_peers_available.clone();
+        seed_peers.shuffle(&mut rand::thread_rng());
+        let mut seed_peers_iter = seed_peers.iter();
+        let mut permitted_concurrent = num_seeds_to_try;
+        let mut remaining_seed_peers = seed_peers.len();
+
+        let mut exit_loop = false;
+        let max_peers_to_sync_per_round = self.config().network_discovery.max_peers_to_sync_per_round;
+        let max_permitted_peer_claims = self.config().max_permitted_peer_claims;
+        let max_permitted_peer_addresses_per_claim = self
+            .config()
+            .peer_validator_config
+            .max_permitted_peer_addresses_per_claim;
+        while !exit_loop && permitted_concurrent > 0 {
```

Comment:
why do this in a while?
the while loop can repeat, but you will only dial the same peers again, which will probably fail/give the same results again.

Reply:
No, new peers are selected each time until the iterator is empty
```
let candidates: Vec<Peer> = seed_peers_iter.by_ref().take(permitted_concurrent).cloned().collect();
```
We dial concurrently, but only up to the number of peers as per `max_peers_to_sync_per_round`, see how many did not respond, then select that number to dial concurrently again until the quota is full.

Reply:
I see